File: src/ObjectManager/Object.Tes/FilePacks/Records/345-ALCH.Potion.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class ALCHRecord : Record, IHaveEDID, IHaveMODL
    {
        // TESX
        public class DATAField
        {
            public float Weight;
            public int Value;
            public int Flags; //: AutoCalc

            public DATAField(UnityBinaryReader r, int dataSize, GameFormatId formatId)
            {
                Weight = r.ReadLESingle();
                if (formatId == GameFormatId.TES3)
                {
                    Value = r.ReadLEInt32();
                    Flags = r.ReadLEInt32();
                }
            }

            public void ENITField(UnityBinaryReader r, int dataSize)
            {
                Value = r.ReadLEInt32();
                Flags = r.ReadByte();
                r.ReadBytes(3); // Unknown
            }
        }

        // TES3
        public struct ENAMField
        {
            public short EffectId;
            public byte SkillId; // for skill related effects, -1/0 otherwise
            public byte AttributeId; // for attribute related effects, -1/0 otherwise
            public int Unknown1;
            public int Unknown2;
            public int Duration;
            public int Magnitude;
            public int Unknown4;

            public ENAMField(UnityBinaryReader r, int dataSize)
            {
                EffectId = r.ReadLEInt16();
                SkillId = r.ReadByte();
                AttributeId = r.ReadByte();
                Unknown1 = r.ReadLEInt32();
                Unknown2 = r.ReadLEInt32();
                Duration = r.ReadLEInt32();
                Magnitude = r.ReadLEInt32();
                Unknown4 = r.ReadLEInt32();
            }
        }

        public override string ToString() => $"ALCH: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model
        public STRVField FULL; // Item Name
        public DATAField DATA; // Alchemy Data
        public ENAMField? ENAM; // Enchantment
        public FILEField ICON; // Icon
        public FMIDField<SCPTRecord>? SCRI; // Script (optional)
        // TES4
        public List<ENCHRecord.EFITField> EFITs = new List<ENCHRecord.EFITField>(); // Effect Data
        public List<ENCHRecord.SCITField> SCITs = new List<ENCHRecord.SCITField>(); // Script Effect Data

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL": if (SCITs.Count == 0) FULL = new STRVField(r, dataSize); else ArrayUtils.Last(SCITs).FULLField(r, dataSize); return true;
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "DATA":
                case "ALDT": DATA = new DATAField(r, dataSize, formatId); return true;
                case "ENAM": ENAM = new ENAMField(r, dataSize); return true;
                case "ICON":
                case "TEXT": ICON = new FILEField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                //
                case "ENIT": DATA.ENITField(r, dataSize); return true;
                case "EFID": r.ReadBytes((int)dataSize); return true;
                case "EFIT": EFITs.Add(new ENCHRecord.EFITField(r, dataSize, formatId)); return true;
                case "SCIT": SCITs.Add(new ENCHRecord.SCITField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-ALCH.Potion.swift
//
//  ALCHRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class ALCHRecord: Record, IHaveEDID, IHaveMODL {
    // TESX
    public class DATAField {
        public let weight: Float
        public var value: Int32 = 0
        public var flags: Int32 = 0 //: AutoCalc

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            weight = r.readLESingle()
            if format == .TES3 {
                value = r.readLEInt32()
                flags = r.readLEInt32()
            }
        }

        func ENITField(_ r: BinaryReader, _ dataSize: Int) {
            value = r.readLEInt32()
            flags = Int32(r.readByte())
            r.skipBytes(3) // Unknown
        }
    }

    // TES3
    public struct ENAMField {
        public let effectId: Int16
        public let skillId: UInt8 // for skill related effects, -1/0 otherwise
        public let attributeId: UInt8 // for attribute related effects, -1/0 otherwise
        public let unknown1: Int32
        public let unknown2: Int32
        public let duration: Int32
        public let magnitude: Int32
        public let unknown4: Int32

        init(_ r: BinaryReader, _ dataSize: Int) {
            effectId = r.readLEInt16()
            skillId = r.readByte()
            attributeId = r.readByte()
            unknown1 = r.readLEInt32()
            unknown2 = r.readLEInt32()
            duration = r.readLEInt32()
            magnitude = r.readLEInt32()
            unknown4 = r.readLEInt32()
        }
    }

    public override var description: String { return "ALCH: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model
    public var FULL: STRVField! // Item Name
    public var DATA: DATAField! // Alchemy Data
    public var ENAM: ENAMField? = nil // Enchantment
    public var ICON: FILEField! // Icon
    public var SCRI: FMIDField<SCPTRecord>? = nil // Script (optional)
    // TES4
    public var EFITs = [ENCHRecord.EFITField]() // Effect Data
    public var SCITs = [ENCHRecord.SCITField]() // Script Effect Data

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL": if SCITs.count == 0 { FULL = r.readSTRV(dataSize) } else { SCITs.last!.FULLField(r, dataSize) }
        case "FNAM": FULL = r.readSTRV(dataSize)
        case "DATA",
             "ALDT": DATA = DATAField(r, dataSize, format)
        case "ENAM": ENAM = ENAMField(r, dataSize)
        case "ICON",
             "TEXT": ICON = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        //
        case "ENIT": DATA.ENITField(r, dataSize)
        case "EFID": r.skipBytes(dataSize)
        case "EFIT": EFITs.append(ENCHRecord.EFITField(r, dataSize, format))
        case "SCIT": SCITs.append(ENCHRecord.SCITField(r, dataSize))
        default: return false
        }
        return true
    }
}
